### PR TITLE
OpenImageIO: conan 2 support, old version cleanup & version 2.4.17.0 & 2.5.6.0 added

### PR DESCRIPTION
**Pull request by irieger**
Specify library name and version:  **openimageio/2.x**

Main change: Conan 2 support.

Add **2.5.6.0** & **2.4.17.0**, remove 2.2.7.0 & 2.2.18.0 & 2.3.7.2.

This pull request is adapting the existing openimageio recipe to Conan 2.0 for the newest version of OpenImageIO (as of starting the MR). I removed some old versions and offering 3 versions inside the 2.* major which I understand the FAQ to be requiring.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by irieger** [comment]
Any hint what is required for the CI to run/why it currently cannot be built? Just got approved and hour ago and repushed with the current master merged into my branch and it failed. Clicking on Details results in a "403 Forbidden" error.

Btw. I saw that my quick fix for OpenColorIO requirements is not needed if #19549 is merged first which also updates OpenColorIO and looks good to me.

Also a small question on the proper workflow here: I'm normally used to projects on non-public gitlab servers where the merge workflow is to have gitlab squash commits for a flat graph. How is that handled here? Squashing the pull request on merge or preparing it before merging?

**Review by uilianries — APPROVED** [review]

File: recipes/openimageio/all/conanfile.py
Comment (EstebanDugueperoux2) [inline recipes/openimageio/all/conanfile.py:246]:
```suggestion
            "imath::imath",
```
To avoid the "ConanException: error generating context for 'openimageio/2.3.7.2': {'ref': Imath/unknown@unknown/unknown, 'build': False, 'direct': True, 'test': False, 'visible': True}" error.

**Review by EstebanDugueperoux2 — CHANGES_REQUESTED** [review]
Hi, I have the same error on my workstation at test_package step, the error is in the package_info().

**Comment by irieger** [comment]
> Hi, I have the same error on my workstation at test_package step, the error is in the package_info().

Hey, thank you for the hint. Happy to have someone look at this too. It is very hard to work on this when your system just compiles fine and the CI fails which is rather hard to find the exact error some time. I think I had it also in lower case and it failed too but lets see what the CI thinks.

File: recipes/openimageio/all/conanfile.py
Comment (EstebanDugueperoux2) [inline recipes/openimageio/all/conanfile.py:218]: In https://c3i.jfrog.io/c3i/misc/logs/pr/19950/38-windows-visual_studio/openimageio/2.3.7.2//b2c07836b65cee5d00045084ae8eafc572e6d078-build.txt i see an error about system dlls in bien folder.

Comment (jmarrec) [reply recipes/openimageio/all/conanfile.py:218]: @irieger Try something like this: https://github.com/conan-io/conan-center-index/blob/a523d487eeb89c3c5cdef54d4a8d15a6d1df4974/recipes/mongo-cxx-driver/all/conanfile.py#L155-L160

File: recipes/openimageio/all/conanfile.py
Comment (jmarrec) [inline recipes/openimageio/all/conanfile.py:215]:
```suggestion
        rmdir(self, os.path.join(self.package_folder, "share"))
        if self.settings.os == "Windows":
            for vc_file in ("concrt", "msvcp", "vcruntime"):
                rm(self, f"{vc_file}*.dll", os.path.join(self.package_folder, "bin"))
```

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py:218]: Yeah, saw that message too. Have nearly no clue about the build on crap OS. However your rmdir seems a bit wide reaching to me without knowing how exactly the folders on Windows might look here. Would hope that pkgconfig and cmake aren't filled with actual dlls? Did you test that on Windows?

Comment (jmarrec) [reply recipes/openimageio/all/conanfile.py:221]: The rmdir isn't necessarilly needed, and your library may not contain them to begin with. I was just showing context (lines above and below).

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py:242]: I oriented myself on the v1 conanfile as well as OpenEXR which has several components. But I'm open to change that.

File: recipes/openimageio/all/test_package/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/test_package/conanfile.py]: Adapted some changes to the conanfile for the package and reverted the v1 package to coexist. Is there anything specific required to wire that?

**Comment by EstebanDugueperoux2** [comment]
It remains a build error about ffmpeg dependency:

`Missing prebuilt package for 'ffmpeg/4.4.4'`

This dependency release is only needed by old "2.3.7.2" openimageio release.
Perhaps we can drop this release to avoid this ffmpeg error.

**Comment by irieger** [comment]
> It remains a build error about ffmpeg dependency:
> 
> `Missing prebuilt package for 'ffmpeg/4.4.4'`
> 
> This dependency release is only needed by old "2.3.7.2" openimageio release. Perhaps we can drop this release to avoid this ffmpeg error.

Yes, I just removed 2.3.7.2. I think it should not be needed and maybe this helps to move forward.

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py]: I did rename both variables that had a bad code style

**Comment by EstebanDugueperoux2** [comment]
> > It remains a build error about ffmpeg dependency:
> > `Missing prebuilt package for 'ffmpeg/4.4.4'`
> > This dependency release is only needed by old "2.3.7.2" openimageio release. Perhaps we can drop this release to avoid this ffmpeg error.
> 
> Yes, I just removed 2.3.7.2. I think it should not be needed and maybe this helps to move forward.

We got yet a "MISSING_DEPENDENCIES" error yet, do you have the same error by using  ffmpeg 6.1 dependency?

**Comment by irieger** [comment]
> > > It remains a build error about ffmpeg dependency:
> > > `Missing prebuilt package for 'ffmpeg/4.4.4'`
> > > This dependency release is only needed by old "2.3.7.2" openimageio release. Perhaps we can drop this release to avoid this ffmpeg error.
> > 
> > 
> > Yes, I just removed 2.3.7.2. I think it should not be needed and maybe this helps to move forward.
> 
> We got yet a "MISSING_DEPENDENCIES" error yet, do you have the same error by using ffmpeg 6.1 dependency?

Just bumped all dependencies. Lets see if that helps.

File: recipes/openimageio/all/conanfile.py
Comment (EstebanDugueperoux2) [inline recipes/openimageio/all/conanfile.py:102]: No luck, now there  is a conflict on libpng releases because of freetype.
Perhaps we could relax the requirement on libpng:
```suggestion
            self.requires("libpng/[>=1.6.40 <2]")
```

Comment (irieger) [reply recipes/openimageio/all/conanfile.py:102]: Not sure it is is only one cause there is also complaints about libtiff (which I didn't touch) https://c3i.jfrog.io/c3i/misc/logs/pr/19950/47-linux-gcc11/openimageio/2.4.7.1//543ed657bef9ccadbcb98df436e25612de3b2d3f-build.txt

So I don't really get how one is supposed to work with this. Just pushing things and waiting for hours in the CI and then push another try for dependencies can't be the solution?!

(To be honest I don't get the prebuild packages concept anyway as I always run into mismatched dependencies and stuff. Also for some stuff I just sometimes anyway want to configure options so prebuild hasn't worked for me for all but quick tests.)

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/1)

**Comment by EstebanDugueperoux2** [comment]
> The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/1)

Hi @ericLemanissier do you thinks "self.requires("libpng/[>=1.6.39 <2]")" should fix these errors?

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/2)

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/3)

**Comment by ericLemanissier** [comment]
@EstebanDugueperoux2 it was a transient error. it seems that conan-center sometimes distribute outdated recipes.
The error remaining on linux is about missing ffmpeg binaries, which is because CI of https://github.com/conan-io/conan-center-index/commit/f0e0a868ddecd683f8f61d6fc1fc26d2806f4ef6 is not yet finished. (but is it still actually running after all this time)

**Comment by EstebanDugueperoux2** [comment]
@irieger  could you close and reopen this PR to trigger again the pipelines?

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/4)

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/5)

**Comment by ericLemanissier** [comment]
The build is now finished. You can take a look at the results [here](https://github.com/ericLemanissier/proof-of-conan/actions/runs/7757809832/attempts/6)

**Review by EstebanDugueperoux2 — APPROVED** [review]

**Comment by EstebanDugueperoux2** [comment]
Hello @danimtb , @lasote @czoido @memsharded ,
we need some review/approval on this MR, could you help us on it?

File: recipes/openimageio/all/conanfile.py
Comment (RubenRBS) [inline recipes/openimageio/all/conanfile.py:163]: Is robinmap a necessary or optional lib? 
The wording in https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/43fea763945991d1b83f5c0bb65f12bfab09b99b/src/cmake/externalpackages.cmake#L300 makes it seem optional, and it currently is vendorized because we seem to lack that library in CCI (tracking in #22872)

File: recipes/openimageio/all/conanfile.py
Comment (RubenRBS) [inline recipes/openimageio/all/conanfile.py:166]: Should we maybe have `tools.build:skip_test` be reflected here?

File: recipes/openimageio/all/conanfile.py
Comment (RubenRBS) [inline recipes/openimageio/all/conanfile.py:207]:
```suggestion
        copy(self, "LICENSE*.md", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
```

We still need to copy the old BSD license

File: recipes/openimageio/all/conanfile.py
Comment (RubenRBS) [inline recipes/openimageio/all/conanfile.py:234]:
```suggestion
```

No longer needed

File: recipes/openimageio/all/test_v1_package/CMakeLists.txt
Comment (RubenRBS) [inline recipes/openimageio/all/test_v1_package/CMakeLists.txt:10]:
```suggestion
target_link_libraries(${PROJECT_NAME} OpenImageIO::OpenImageIO)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_14)
```

File: recipes/openimageio/all/test_package/CMakeLists.txt
Comment (RubenRBS) [inline recipes/openimageio/all/test_package/CMakeLists.txt:10]:
```suggestion
                                              fmt::fmt)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_14)
```

File: recipes/openimageio/all/test_v1_package/conanfile.py
Comment (RubenRBS) [inline recipes/openimageio/all/test_v1_package/conanfile.py:11]:
```suggestion
        cmake = CMake(self)
```

No longer needed now that 2.4.7 is the oldest one available

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py:166]: Maybe. I'm not fully sure how that works. I just did a best guess by looking at other related projects like openexr and the old conan 1 file. How would that best be done?

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py:163]: To my understanding it is needed and the code you linked allows between selecting an external one vs. having Cmake download a robinmap and build it itself. But maybe I'm missunderstanding something.

Comment (irieger) [reply recipes/openimageio/all/conanfile.py:163]: Sorry, misunderstood that earlier. To my understanding it was by the changes and patches using tsl-robin-map as the robinmap. Not sure what the difference is if any to the MR you linked.

Comment (irieger) [reply recipes/openimageio/all/conanfile.py:161]: So, testing is no fixed to false.

File: recipes/openimageio/all/conanfile.py
Comment (irieger) [reply recipes/openimageio/all/conanfile.py]: Removed option and fixed to false

File: recipes/openimageio/all/test_v1_package/CMakeLists.txt
Comment (RubenRBS) [inline recipes/openimageio/all/test_v1_package/CMakeLists.txt:11]:
```suggestion
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
```
I'm a dum dum, sorry!

File: recipes/openimageio/all/test_package/CMakeLists.txt
Comment (RubenRBS) [inline recipes/openimageio/all/test_package/CMakeLists.txt:11]:
```suggestion
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
```

**Review by RubenRBS — COMMENTED** [review]
Oops

**Review by EstebanDugueperoux2 — APPROVED** [review]

**Review by uilianries — COMMENTED** [review]
LGTM. I'll just wait for @RubenRBS  final review

**Review by uilianries — APPROVED** [review]

**Review by RubenRBS — APPROVED** [review]

File: recipes/openimageio/all/patches/2.5.6.0-cmake-targets.patch
Comment (lgritz) [inline recipes/openimageio/all/patches/2.5.6.0-cmake-targets.patch:121]: Just FYI, the checked_find_package macro itself internally checks USE_Package/USE_PACKAGE and disables if either exists and is set to 0. So I'm not sure that it was necessary to put the `if (USE_PACKAGE)` around all these various dependency calls.

Comment (irieger) [reply recipes/openimageio/all/patches/2.5.6.0-cmake-targets.patch:121]: To be hones, I'm not sure, I just followed the existing patches for older versions.

**Pull request merged**
